### ISSUE TITLE
gpt-j input shape after finetuning

### ISSUE THREAD

**Issue by calderma** [huggingface/transformers#13581]
Hello after finetuning the EleutherAI/gpt-j-6B model and loading one of the checkpoints I'm getting the following error:

size mismatch for transformer.wte.weight: copying a param with shape torch.Size([50257, 4096]) from checkpoint, the shape in current model is torch.Size([50400, 4096])

Based on the GPT-J description there are some extra tokens beyond what the base GPT2 tokenizer has. Running the normal EleutherAI/gpt-j-6B model and tokenizer works fine, so perhaps the config that's being saved with the checkpoint is not accounting for those extra tokens? I pulled master this morning so I believe this is still the case with the current codebase. Training script is:
deepspeed run_clm.py \
    --model_name_or_path EleutherAI/gpt-j-6B \
    --train_file $TRAIN_FILE \
    --validation_file $VAL_FILE \
    --gradient_accumulation_steps 8 \
    --num_train_epochs 2.0 \
    --evaluation_strategy epoch \
    --save_steps XXX \
    --logging_steps XXX \
    --do_train \
    --do_eval \
    --save_total_limit 3 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --deepspeed "deepspeed_zero3_config.json" \
    --overwrite_output_dir \
    --output_dir gpt-j_model

test script breaks at model loading:

from transformers import GPTJForCausalLM, AutoTokenizer
import torch
from pytictoc import TicToc

t = TicToc()

device = "cuda:0" if torch.cuda.is_available() else "cpu"

model = GPTJForCausalLM.from_pretrained("gpt-j_end2end/checkpoint-XXX", torch_dtype=torch.float16,max_length=2048).to(device)
print("model loaded")
...

Any help would be appreciated. Thanks!

**Comment by patil-suraj** [comment]
Hi @calderma 

Thank you for opening the issue. The actual vocab size is 50257  the remaining are extra tokens. 
The issue is that the vocab size is `50400` in the `config` file, but the tokenizer size is 50257, and the `run_clm.py` script, resizes  token embeddings here 
https://github.com/huggingface/transformers/blob/421929b556aedf022a1c4a1f3b2f116b14a7b88a/examples/pytorch/language-modeling/run_clm.py#L358

which reduces the model's embedding size to 50257, hence the shape mismatch.

@LysandreJik  should we remove the extra tokens in the embedding and update the model on the hub?

**Comment by calderma** [comment]
Ok is there a way around this issue by manually changing the config.json or something or should I just wait until this gets updated on the hub?

**Comment by patil-suraj** [comment]
Yes, if you have already fine-tuned a model, then you should change the `vocab_size` in `config.json` to 50257.

The fix is on the way #13617

**Comment by MantasLukauskas** [comment]
@patil-suraj but in fine-tuned model config.json file vocab_size is already 50257 :)

**Comment by calderma** [comment]
yes I noticed that as well. I got it to run by doing:

from transformers import GPTNeoForCausalLM, AutoTokenizer, GPT2Tokenizer, GPTJForCausalLM
import torch
from pytictoc import TicToc
t = TicToc()

device = "cuda:0" if torch.cuda.is_available() else "cpu"

model = GPTJForCausalLM.from_pretrained("gpt-j_model",ignore_mismatched_sizes=True,torch_dtype=torch.float16).to(device)

print("model loaded")

tokenizer = AutoTokenizer.from_pretrained("gpt2",truncation=True)
model.resize_token_embeddings(len(tokenizer))

but then it output nonesense so I don't think the tokenization is being handled correctly when I do that.

**Comment by patil-suraj** [comment]
> @patil-suraj but in fine-tuned model config.json file vocab_size is already 50257 :)

If that is the case, then the model should load without any issues, as the shapes would match.

**Comment by working12** [comment]
@calderma How exactly are you finetuning? Can you share a complete example somehow?

**Comment by dunalduck0** [comment]
@patil-suraj what's the status of this issue? I looked at your PR but it has no code change.

I am blocked by the same issue when NOT use --fp16. I tried two fine-tune with the same parameters but

(1) one with "--model_revision float16 --fp16"
(2) the other with "--model_revision float32" (no --fp16).

The output model from (1) can be loaded fine but (2) triggered size mismatch error. How do I fix the error? --fp16 fine tuning has so many OVERFLOW errors that I wanted to turn --fp16 off.

**Comment by patil-suraj** [comment]
@dunalduck0  the `main` branch has the fp32 weights which should be used instead of the `fp16` branch.
Also, there was another issue with resizing embeddings, #14190 should fix that.

**Comment by dunalduck0** [comment]
Thank you @patil-suraj

**Comment by patil-suraj** [comment]
This should have been fixed by #14190

**Comment by dunalduck0** [comment]
@alexorona I've got the same problem as you. After resizing, I cannot load the original model anymore due to size mismatch. I found this is due to resize_token_embeddings.() changing config.vocal_size. If I reset config.vocab_size to 54000 before loading, it works fine.

> @patil-suraj @calderma @MantasLukauskas The issue is actually more fundamental. There seems to be a problem with the serialization of GPTJ. Even without fine-tuning, this issue is encountered if the embedding size of GPTJ is expanded. For example, if you run this code below you will also get a size mismatch error:
> 
> ```
> from transformers import GPTJForCausalLM
> model = GPTJForCausalLM.from_pretrained("EleutherAI/gpt-j-6B")
> model.resize_token_embeddings(54001)
> model.save_pretrained("model_dir")
> model = GPTJForCausalLM.from_pretrained("model_dir")
> RuntimeError: Error(s) in loading state_dict for GPTJForCausalLM:
> 	size mismatch for lm_head.weight: copying a param with shape torch.Size([50400, 4096]) from checkpoint, the shape in current model is torch.Size([54001, 4096]).
> 	size mismatch for lm_head.bias: copying a param with shape torch.Size([50400]) from checkpoint, the shape in current model is torch.Size([54001]).
> ```
